packages/cosmic-swingset/src/import-kernel-db.js

```diff
@@ -0,0 +1,205 @@
+#! /usr/bin/env node
+
+// @ts-check
+import '@endo/init';
+
+import os from 'os';
+import process from 'process';
+import { Transform } from 'stream';
+import fsPower from 'fs';
+import fsPromisesPower from 'fs/promises';
+import pathPower from 'path';
+
+import BufferLineTransform from '@agoric/internal/src/node/buffer-line-transform.js';
+import { Fail } from '@agoric/assert';
+import { importSwingStore } from '@agoric/swing-store';
+
+import { isEntrypoint } from './helpers/is-entrypoint.js';
+import { makeProcessValue } from './helpers/process-value.js';
+
+/**
+ * @typedef {object} StateSyncImporterOptions
+ * @property {string} stateDir the directory containing the SwingStore to export
+ * @property {string} exportDir the directory where to place the exported artifacts and manifest
+ * @property {number} [blockHeight] block height to check for
+ * @property {boolean} [includeHistorical] whether to include historical artifacts in the export
+ */
+
+/**
+ * @param {StateSyncImporterOptions} options
+ * @param {object} powers
+ * @param {Pick<import('fs/promises'), 'readFile'> & Pick<import('fs'), 'createReadStream'>} powers.fs
+ * @param {import('path')['resolve']} powers.pathResolve
+ * @param {typeof import('@agoric/swing-store')['importSwingStore']} [powers.importSwingStore]
+ * @param {(...args: any[]) => void} [powers.log]
+ * @returns {Promise<void>}
+ */
+export const performStateSyncImport = async (
+  { stateDir, exportDir, blockHeight, includeHistorical },
+  {
+    fs: { createReadStream, readFile },
+    pathResolve,
+    importSwingStore: importDB = importSwingStore,
+    log,
+  },
+) => {
+  /** @param {string} allegedRelativeFilename */
+  const safeExportFileResolve = allegedRelativeFilename => {
+    const resolvedPath = pathResolve(exportDir, allegedRelativeFilename);
+    resolvedPath.startsWith(exportDir) ||
+      Fail`Exported file ${allegedRelativeFilename} must be in export dir ${exportDir}`;
+    return resolvedPath;
+  };
+
+  const manifestPath = safeExportFileResolve('export-manifest.json');
+  /** @type {Readonly<import('./export-kernel-db.js').StateSyncManifest>} */
+  const manifest = await readFile(manifestPath, { encoding: 'utf-8' }).then(
+    data => JSON.parse(data),
+  );
+
+  if (blockHeight !== undefined && manifest.blockHeight !== blockHeight) {
+    Fail`State-sync manifest for unexpected block height ${manifest.blockHeight} (expected ${blockHeight})`;
+  }
+
+  if (!manifest.data) {
+    throw Fail`State-sync manifest missing export data`;
+  }
+
+  if (!manifest.artifacts) {
+    throw Fail`State-sync manifest missing required artifacts`;
+  }
+
+  const artifacts = harden(Object.fromEntries(manifest.artifacts));
+
+  if (
+    includeHistorical &&
+    manifest.mode !== 'archival' &&
+    manifest.mode !== 'debug'
+  ) {
+    throw Fail`State-sync manifest missing historical artifacts`;
+  }
+
+  /** @type {import('@agoric/swing-store').SwingStoreExporter} */
+  const exporter = harden({
```

Comment:
Suggestion: clarify the somewhat surprising definition of an `exporter` variable inside the function implementing import.
```suggestion
  // Represent the data in `exportDir` as a SwingSetExporter object.
  /** @type {import('@agoric/swing-store').SwingStoreExporter} */
  const exporter = harden({
```